Merge in dev for 0.3.0 Virtac release

ptsOSL opened the pull request:
-Major refactor to PV creation and data management
-Reworked how many PVs work, how the simulation updates PVs and how the PVs update the simulation
-Reworked and updated the docs
-Type hinting and docs strings
-Modified the CSV files to support the refactor
-Streamlined the csv creation process.

ptsOSL committed:
Make waveform val arrays consistent across csv files

We set the default numpy printing options to be large so that large
arrays are correctly printed to csv files without summarization or
newlines.

ptsOSL committed:
Improve tunefb csv file headings

ptsOSL committed:
Add PVs attached to lattice to limits csv file

Previously we were only getting PVs connected to lattice elements
and not the lattice itself. I also added a check that the field is
available in both the simulated lattice and live lattice as some of the
lattices simulated fields do not have live equivalents to caget

ptsOSL committed:
Rename csv headings and add scan column

Mirror.csv now has a scan column to allow users to configure a PV
to process at one of the standard epics rates. This functionality
essentially replaces the old refresh records which will be removed
in a later commit

ptsOSL committed:
Remove refresh records

The 3 refresh PVs are not needed, instead we set the
SCAN field for the 3 input PVs defined for the refresh PVs
to be on a 1 second polling interval

ptsOSL committed:
Check that a new PV is not already in the csv files

This avoids duplicate PVs, as there are some pytac elements which
share the same PV in the LIVE machine. This was harmless, but it
is better not to have redundant data is it is confusing. Also small
type hinting improvement

ptsOSL committed:
Add scan col to limits.csv and update heading names

Also use the new scan column for both elements of the lattice and the
lattice itself

ptsOSL committed:
Add latest generated csv files

ptsOSL committed:
Improve type hinting

ptsOSL committed:
Improve use of numpy printoptions

Our selected print options are now only used within the with statement.

ptsOSL committed:
Specify epicscorelibs version to avoid issue in library

ptsOSL committed:
Refactor generate_pv_limits and do cagets async

I split generate_pv_limits into two functions to allow us to spawn a
cothread for each pytac element. This cothread will asyncronously do
a caget to get the PV data for this element.

ptsOSL committed:
Remove python3.13 from ci and toml file

Due to current incompatibility with epicscorelibs. The default python
version has also changed to a more conservative 3.10

ptsOSL committed:
Added PV class

Most functionality in this file is refactored code from
virtac_main.

ptsOSL committed:
Remove unused python files and update __init__.py

Functionality from masks and mirror_objects is now in pv.py.
The mirror objects are now Inversion, Summation and Collation PV subclasses.
The functionality in the masks is mostly in the RefreshPV

ptsOSL committed:
Rename enable_emittance and add enable_tunefb bool

Instead of using a disable_emmitance flag with default false,
we use enable_emittance with default true, as I think this
makes more sense. I have also introduced an equivelant for
tunefb and the tunefb setup is not initiated from within VirtacServer
rather than __main__.py

ptsOSL committed:
Add method to print info about virtac

This is called at init to give more context instead
of the old print statement

ptsOSL committed:
Move on_update method to PV classes

ptsOSL committed:
Minor code comment and print statement changes

ptsOSL committed:
Update some docstrings and type hinting

ptsOSL committed:
Implement PV classes for pv configuration

This large commit required:
-Importing the new pv classes
-Creating softioc records using the new pv classes, softioc.builder is
no longer required in virtac_server
-Refactoring/improving the handling of some of the csv configuration data
-Refactoring the method which configures "normal" pvs. This has been
replaced by 3 functions, one for making lattice pvs, one for element pvs
and one to call these two methods
-Removal of the 9 VirtacServer owned dict/lists which stored the various
softioc record types, there is now two dicts, one which holds all PVs
and one which holds the read only PVs
-Refactoring of _create_mirror_records, required to use the PV classes,
some of this code has moved to pv.py

ptsOSL committed:
Refactor method to enable/disable PV monitoring

monitor_mirrored_pvs no longer needs to be called explicitly, monitoring
of mirror pvs and mask pvs is now done automatically. With the mirror
functionality done by CollationPV, InversionPV and SummationPV. The
tunefb monitoring which was done through masks.py is now done with the
RefreshPV subclass. However the user can explicitly enable/disable
monitoring which enables/disables all of this behaviour after init.

ptsOSL committed:
Add checks to RecordData for invalid dtypes

ptsOSL committed:
Also check if the monitored PV is a CaPV

ptsOSL committed:
Pass a valid default dtype for SCAN

ptsOSL committed:
Make the virtac print virtac stats by default

ptsOSL committed:
Refactor ca config into its own func

ptsOSL committed:
Bump python and atip versions

We require atip>=0.2 as this is when we split atip/virtac
atip 0.2 requires python3.11, so I bumped that too

ptsOSL committed:
Remove python3.10 and add py3.13 to CI

ptsOSL committed:
Rework how MonitorPV is configured

MonitorPV now only takes a list of strings to initialise
PV monitoring. The classes which inherit MonitorPV still take a PV
as their arguments, but pass pv.name to the super() call. This makes
the MonitorPV class a little cleaner.

ptsOSL committed:
Refactored MonitorPV setup_pv_monitoring

I have made the differences between passing a single callback
function and multiple callbacks more explicit by splitting it
into two seperate functions.

ptsOSL committed:
Switch on tunefb by default and rename args

Tunefb and emmittance are both on by default. I have
also changed both args to be disable_ rather than enable_
. It is more consistent now and consistent with ATIP. It also
makes more sense this way around as they are on by default.

ptsOSL committed:
Fix type checker errors

ptsOSL committed:
No longer fail docs build on warning

Previously the docs build would fail when a sphinx nitpicky
warning occurred. These are extremely common and hard to
fix without just ignoring them elsewhere. There are multiple
issues with nitpicky such as https://github.com/sphinx-doc/sphinx/issues/10785
. I dont think it is worth the effort fixing all of these
nitpicky issues right now, so I have decided to ignore them,
the only effect will be that some links in the API docs
wont work.

ptsOSL committed:
Remove deleted files from __init__.py

ptsOSL committed:
Fix some docstrings, logging and exception strings

ptsOSL committed:
Remove redundant Interim PV

The InversionPV can now invert a list of ai records, so
a seperate PV which does this functionality is no longer
needed

ptsOSL committed:
Split waveform records between in & out waveforms

When we set a value to a waveformOut, it will not automatically
proc, so this was being done explicitly in some cases. Now we
correctly set records to be either in or out rather than all to out
which has allowed me to remove most procs. The bba waveforms should be
waveformOut as they are updated by an external application. The
waveforms which store all bpm possitions should be in as they arent set
externally and need to proc when they change.

ptsOSL committed:
Correct SCAN types

aIn and waveformIn should be SCAN=I/O Inter, aOut
and waveformOut should be SCAN=Passive by default

ptsOSL committed:
Update logging and docstrings

ptsOSL committed:
Fix bug with setup_pv_monitoring

I realised that when re-enabling monitoring, because of
the way we store the data in _monitor_data, it was not
possible to setup the monitoring in the same way as it
would have been done originally. ie if the user setup a
group monitoring, after disabling/enabling monitoring it
would have become multiple single monitors.

ptsOSL committed:
Remove redundant issubclass checks

ptsOSL committed:
Replace toggle monitoring with enable/disable

To avoid redefining the meaning of toggling from
ATIP

ptsOSL committed:
Improve typing and docstrings

ptsOSL committed:
Remove unused CaPV

ptsOSL committed:
Fix CollatePV and InvertPV callbacks

This is much nicer now. When the Invert/Collate callback
is called, previously we did a get on every monitored pv,
now we just update a single value using the index and value
supplied. It should reduce computation time a little

ptsOSL committed:
Move where the pytac lattice is set

Rather than updating the lattice in the on_update callback
of the OUT pv, we update it in the set function of the
IN pv. This makes more sense logically. The default PV
now sets its attached lattice element if it exists, the
ReadbackPV does not, so overwrites the set method

ptsOSL committed:
Remove pypi from copier config and CI

Not needed for this repo

ptsOSL committed:
Implement enums for record types and mirror PVs

ptsOSL committed:
Reduce code in print_virtac_stats

ptsOSL committed:
Change PV baseclass and move pytac attributes to new subclass

We decided it makes more sense to put the pytac elements/fields in
a subclass. To create a better seperation between PVs which read
from the simulation and PVs which write to it, I also created
another subclass, ReadWriteSimPV.

ptsOSL committed:
Merge Offset and Setpoint classes into ProxyPV

It was decided that having two classes was overkill, the
name also better reflects its purpose. Additionally some
attribute names were changed for clarity

ptsOSL committed:
Change MonitorPV and subclasses for clarity and improved index handling

All callbacks are now called _callback for clarity, they
all now overwrite the base _callback method. Handling of when index
is None has also been improved for cases where index is None is
not currently passed.

ptsOSL committed:
Slim down core PV creation using new PV class changes

New changes to the PV classes has allowed me to remove most uses of the
functions to attach pytac data. It has also allowed a better seperation
between where PVs which write to pytac and PVs which read from pytac are
created.

ptsOSL committed:
Remove most code from update_pvs

Now we just do a get on the ReadSimPV which gets the
data from the simulation and puts it in the softioc record

ptsOSL committed:
Misc docstring improvements and fixes from PV class changes

ptsOSL committed:
Remove unused function: set_feedback_record

ptsOSL committed:
Update logging

ptsOSL committed:
Remove ProxyPV

The ReadWriteSimPV, which is the SETI PV written to by
HLA apps, now directly updates the pytac lattice, this is more logical
than previously where it was a proxy which made the ReadSimPV update
the pytac lattice. When the ReadWriteSimPV updates, it calls its own
set function which both updates the lattice and then copies the new
value to its linked ReadSimPV. Ideally this would be read back from
the simulation but this has a time cost and currently would just
return the same value.

ptsOSL committed:
Remove python 3.10 and fix min atip version

Also add ci tests for python 3.13

ptsOSL committed:
Fix CLI bugs

Fixed outdated help message. Also fixed a bug caused
by not passing a value in to -v which would default to None.
This would cause an exception when converting to an int. We
now default to 0 and require the passed value to be an int, otherwise
an exception is raised

ptsOSL committed:
Fix bug when enabling/disabling monitoring

We would not clear the monitor list after disabling which would
lead to us duplicating camonitors. I have fixed this behvaiour and
the code which checked for duplicate camonitoring which was broken

ptsOSL committed:
Refactor code for creating element pvs

I have moved things around to improve readability. This was
possible due to previous changes to the PV classes

ptsOSL committed:
Renamed variables in tunefb setup for readability

ptsOSL committed:
Updates to docstrings and log messages from code review

ptsOSL committed:
Use existing pv name list for creating MonitorPVs

ptsOSL committed:
Removed unneeded func to get pytac data

No longer needed as only the PV which owns the data
, now needs to get the data as originally intended

ptsOSL committed:
Make verbosity more standard

Also reordered a function to improve readability

ptsOSL committed:
Merge branch 'dev' into virtac-refactor-rebased

ptsOSL committed:
Merge pull request #8 from DiamondLightSource/virtac-refactor-rebased

Major refactor which replaces the old way that we configured the VIRTACs softioc records and general functionality with a new system of inherited PV classes.

Each PV subclass holds a single softioc record, and all of the attributes and methods required to control it. This behaviour was previously spread throughout virtac_server, masks.py and mirror_objects.py, but is now centralised within the PV classes, with the PV subclasses containing the more complex PV behaviour.

ptsOSL committed:
Update DIAD csv files

I had forgotten to rebuild these after recent changes to create_csv.py

ptsOSL committed:
Set the default data source to pytac.SIM

I dont know why this wasnt already been done, but now it
means that when using commands from the python terminal such
as server.lattice.get_value('emittance'), it looks at the
simulated fields rather than the live ones.

ptsOSL committed:
Change how we import our own modules

Now __main__ is part of __all__

ptsOSL committed:
Fix for incompatible pytac version

ptsOSL committed:
Add create_csv as entrypoint

ptsOSL committed:
Improve docstrings and rename some variables for clarity

ptsOSL committed:
Make github pages update from the docs branch

ptsOSL committed:
Rework the docs in tutorials

I removed the old virtac_example and distributed the content
between overview.md and using_virtac.md

ptsOSL committed:
Add the ability to print the output of commands to docs

Currently used to automatically print the output from
virtac -h and create_csv -h

ptsOSL committed:
Add a file explaining how to update the csv file

ptsOSL committed:
Add a proper main function to create_csv.py

ptsOSL committed:
Minor documentation improvements

ptsOSL committed:
Improve docs, adding detail and clarifications

ptsOSL committed:
Make docs build from dev branch instead of docs

Building it from a docs branch could be confusing

ptsOSL committed:
Update README.md

ptsOSL committed:
Merge pull request #10 from DiamondLightSource/docs

This merge updates all docstrings and the github pages docs so that they are up to date and in a healthy state.
-Rework github pages content to add more information and make it up to date
-Fix various broken docstrings
-Rename a few variables in the source code

ptsOSL committed:
Update to latest copier template

ptsOSL committed:
Clarify use of dev branch

ptsOSL committed:
Fix function return type hinting

Mostly adding -> None for functions which dont
return anything

ptsOSL committed:
Add type casting in various places

For improved readability

ptsOSL committed:
Created CallbackType for improved type hinting

This replaces the more generic Callable type which
was previously expected when passing a callback function to
camonitor

ptsOSL committed:
Improve limits_dict type hinting and convert to str

The data read from the csv files is not manipulated,
it is only used to set epics record fields which require
str type, so we ensure that is the case.

ptsOSL committed:
Correct type hinting from str to pathlib.Path

ptsOSL committed:
Remove unnecessary type assertions

These are not needed as we now have proper type
hinting. (assuming the type hints are followed)

ptsOSL committed:
Check if variable exists instead of is not none

This is to better adhere to type hinting rules

ptsOSL committed:
Fix bug in InversionPVs callback

We were not setting the value if the index was None

ptsOSL committed:
Remove inaccessible check for tune_csv being None

ptsOSL committed:
Pass an element list instead of a single element to ReadSimPV

ptsOSL committed:
Remaining misc type hint improvements

ptsOSL committed:
Remove unneeded commas to improve readability

ptsOSL committed:
Merge pull request #12 from DiamondLightSource/type-hinting-fixes

-Updated to latest copier template
-Switched on pyright strict checking and fixed as many issues as I could
-Switched back to mypy which is now passing and checking all functions
-Spotted and fixed a couple of minor bugs as well
-Removed some code which is redundant assuming the type hinting is followed

MJGaughran commented on src/virtac/pv.py:
```diff
@@ -0,0 +1,607 @@
+        Args:
+            record_data (RecordData): Dataclass used to create this PVs softioc record.
+        """
+        if "self._record" in locals():
```
Perhaps this variable should be properly type hinted and initialised in the init?

MJGaughran replied:
I have never seen this way of checking for an attribute before.

MJGaughran commented on src/virtac/pv.py:
```diff
@@ -0,0 +1,607 @@
+
+
+    Attributes:
+        self.name (str): The name used to get both the PV and its softioc record.
```
Here, and elsewhere: is it possible to remove the duplicate type hints? I can't think of a good reason for keeping them.

ptsOSL replied:
Its a bit subjective this, so ill make some changes and then we can discuss how the docs look after. There is definitely too much fluff in there the API docs atm.

MJGaughran reviewed (commented):
A couple minor issues, but we can cover them in a separate PR if they're non-trivial.

ptsOSL committed:
Remove dupe docstring type hints

Also various minor docstring fixes including fixing returns docstrings.

ptsOSL committed:
Add getter and setter for _record

This is to better handle cases where _record is None
and shouldnt be, and also where it is not None when it should be.
We need a if not None guard in the setter and create_softioc_record as
the setter guard wont be checked until after the record has been made
which is too late.

ptsOSL committed:
Rename the PV class set/get funcs

get_value and set_value are a bit clearer to avoid anyone
thinking that get/set refer to the softioc record rather
than the records value field
